mount-ns.sh

```diff
@@ -0,0 +1,105 @@
+#!/bin/bash
+
+set -eu
+
+# FIXME: This should be replace by a mount of devtmpfs when
+# it is supported in user namespaces
+bind_dev() {
+    dev="${1}/dev"
+
+    for device in null zero full random urandom tty; do
+        touch "${dev}/${device}"
+        mount --bind "/dev/${device}" "${dev}/${device}"
+    done
+}
+
+# Because systemd debian package post-install script is not very
+# robust, we have to create a symlink and bind mount resolv.conf
+# there.
+bind_resolv() {
+    sysroot="${1}"
+
+    resolv_real_path="${sysroot}/run/fake-resolv.conf"
+    create_symlink=yes
+    if [ -L "${sysroot}/etc/resolv.conf" ]; then
+        resolv="$(readlink "${sysroot}/etc/resolv.conf")"
+        if [ "${resolv}" = "../run/systemd/resolve/stub-resolv.conf" ]; then
+            resolv_real_path="${sysroot}/run/systemd/resolve/stub-resolv.conf"
+            create_symlink=no
+        fi
+    fi
+    mkdir -p "$(dirname "${resolv_real_path}")"
+    touch "${resolv_real_path}"
+    mount --bind -o ro /etc/resolv.conf "${resolv_real_path}"
+    if [ "${create_symlink}" = yes ]; then
+        ln -srf "${resolv_real_path}" "${sysroot}/etc/resolv.conf"
+    fi
+}
+
+if [ $# -lt 3 ]; then
+    echo "Expected at least 3 arguments" 1>&2
+    exit 1
+fi
+
+command="${1}"
+sysroot="${2}"
+shift 2
+
+case "${command}" in
+    spawn)
+        cleanup() {
+            umount "${sysroot}/dev" || true
+            umount "${sysroot}/tmp" || true
+            umount "${sysroot}/run" || true
+        }
+        mount -t tmpfs -o mode=0755 tmpfs "${sysroot}/dev"
+        mount -t tmpfs -o mode=1777 tmpfs "${sysroot}/tmp"
+        mount -t tmpfs -o mode=0755 tmpfs "${sysroot}/run"
+        trap cleanup EXIT
+        unshare --pid --fork --mount -- "${0}" init "${sysroot}" "${@}"
```

Comment:
Please add comment clarifying that you are calling the script again with a different command.

Reply:
Done.

I have also noticed there was a potential race condition with the tmpfs when launch multiple time in the same time. So I have changed the tmpfs to be mounted in a temporary directory and bind mounted in the second phase.